spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java

```diff
@@ -333,6 +352,14 @@ private Object doSendAndReceive(Object object, boolean shouldConvert) {
 					error = ((ErrorMessage) reply).getPayload();
 				}
 			}
+			if (reply == null && this.errorOnTimeout) {
```

Comment:
Hi @garyrussell 
We have recently come upon the same issue for migrating to the latest Spring version for using `service-activator` with a `void` return.
Below is how our gateway is currently defined for this flow:
```
<int-http:inbound-gateway path="/v1/fooService/fireEvent"
                          supported-methods="POST"
                          request-channel=“channel.fooService.fireEvent.request"
                          reply-channel=“custom.httpReplyChannel"
                          request-payload-type=“com.fooService.Event"
                          header-mapper=“custom.defaultHttpHeaderMapper"
                          error-channel="custom.httpErrorInputChannel">
    <int-http:request-mapping consumes="application/json" produces="application/json" />
</int-http:inbound-gateway>

<int:service-activator input-channel=“channel.fooService.fireEvent.request"
                       output-channel=“channel.fooService.fireEvent.response"
                       expression="@'com.fooService'.fireEvent(payload)" />
<int:bridge input-channel="channel.fooService.fireEvent.response"
            output-channel=“custom.httpPreReplyChannel" />

<int:header-enricher input-channel="custom.httpPreReplyChannel"
                     output-channel=“custom.httpReplyChannel"
    <int:header name="Set-Cookie" ref="com.fooService.setCookieHeaderEnricher" method="refreshCookieExpiry" />
</int:header-enricher>
```

Currently once the `service-activator` has fired it will send `null` to the `response` channel where we use `<int:header-enricher>` before routing to the `reply` channel on the `http-inbound-gateway`. 
Given that we are using this for a `void` method it would make sense for us to be using `<http:inbound-channel-adapter/>` instead. However our requirements specify that we need specific headers on the response and so have to perform header-enrichment.

Do you know if there is a way that we could perform this using the `<http:inbound-channel-adapter/>`?